nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutEmail.java

```diff
@@ -299,4 +303,55 @@ public void testOutgoingMessageWithFlowfileContent() throws Exception {
         assertEquals("anotherbcc@apache.org", message.getRecipients(RecipientType.BCC)[1].toString());
     }
 
-}
+    @Test
+    public void testUnrecognizedCharset() {
+        runner.setProperty(PutEmail.SMTP_HOSTNAME, "smtp-host");
+        runner.setProperty(PutEmail.HEADER_XMAILER, "TestingNiFi");
+        runner.setProperty(PutEmail.FROM, "test@apache.org");
+        runner.setProperty(PutEmail.MESSAGE, "test message");
+        runner.setProperty(PutEmail.TO, "recipient@apache.org");
+
+        // not one of the recognized charsets
+        runner.setProperty(PutEmail.INPUT_CHARACTER_SET, "NOT A CHARACTER SET");
+
+        runner.assertNotValid();
+    }
+
+    @Test
+    public void testPutEmailWithMismatchedCharset() throws Exception {
+        // String specifically chosen to have characters encoded differently in US_ASCII and UTF_8
+        final String rawString = "SoftwÄrë Ënginëër Ön NiFi";
+        final byte[] rawBytes = rawString.getBytes(StandardCharsets.US_ASCII);
+        final byte[] rawBytesUTF8 = rawString.getBytes(StandardCharsets.UTF_8);
+
+        // verify that the message bytes are different (some messages are not)
+        assertNotEquals(rawBytes, rawBytesUTF8);
+
+        runner.setProperty(PutEmail.SMTP_HOSTNAME, "smtp-host");
+        runner.setProperty(PutEmail.HEADER_XMAILER, "TestingNiFi");
+        runner.setProperty(PutEmail.FROM, "test@apache.org");
+        runner.setProperty(PutEmail.MESSAGE, new String(rawBytes, StandardCharsets.US_ASCII));
+        runner.setProperty(PutEmail.TO, "recipient@apache.org");
+        runner.setProperty(PutEmail.INPUT_CHARACTER_SET, StandardCharsets.UTF_8.name());
+
+        runner.enqueue("Some Text".getBytes());
+
+        runner.run();
+
+        runner.assertQueueEmpty();
+        runner.assertAllFlowFilesTransferred(PutEmail.REL_SUCCESS);
+
+        // Verify that the Message was populated correctly
+        assertEquals("Expected a single message to be sent", 1, processor.getMessages().size());
+        Message message = processor.getMessages().get(0);
+        assertNotEquals(rawString, message.getContent());
```

Comment:
Is it reasonable to tighten up this check?  What exactly are we expecting the differences to be?